File: code_samples/section7/problems/problem7_4/problem7_4.py
"""
Problem 7.4 - Subarray Sum Equals K.

This module implements an efficient solution to count how many contiguous
subarrays sum to a given value k, using prefix sums and a hash table.
"""

def subarray_sum(nums: list[int], k: int) -> int:
    """
    Counts the number of contiguous subarrays whose sum equals k.

    Definition:
        A subarray is a contiguous slice of the array. The task is to count
        how many such slices have a total sum exactly equal to k.

    Args:
        nums:
            List of integers (may be empty and may contain negative values).
        k:
            Target sum.

    Returns:
        Number of contiguous subarrays whose sum is exactly k.

    Core idea (prefix sums + frequency map):
        Let prefix be the running sum of elements seen so far.

        The sum of a subarray from index (j+1) to i is:
            prefix[i] - prefix[j]

        We want:
            prefix[i] - prefix[j] == k
        Rearranging:
            prefix[j] == prefix[i] - k

        So, for each new prefix sum:
          - Count how many times (prefix - k) has appeared before.
          - Add that count to the answer.
          - Record the current prefix sum in a frequency dictionary.

    Why freq starts with {0: 1}:
        - This represents an "empty prefix" before reading any elements.
        - It allows subarrays that start at index 0 to be counted correctly
          when prefix == k.

    Complexity (expected / average):
        - Time:  O(n)
            Single pass through the array; dictionary lookups are O(1) on average.
        - Space: O(n)
            In the worst case, all prefix sums are distinct.

    Notes:
        - Works correctly with negative numbers (unlike sliding window approaches).
        - Python integers do not overflow, so no special handling is required.
    """
    count = 0           # Total number of valid subarrays found
    prefix = 0          # Running prefix sum

    # Frequency map: prefix_sum -> number of times it has appeared so far
    freq: dict[int, int] = {0: 1}

    # Process each element in the array
    for x in nums:
        # Update running sum
        prefix += x

        # We are looking for prior prefix sums equal to (prefix - k)
        need = prefix - k

        # Add how many times that needed prefix sum has appeared
        count += freq.get(need, 0)

        # Record that we've now seen this prefix sum one more time
        freq[prefix] = freq.get(prefix, 0) + 1

    return count


# ============================
# Tests (top-level statements)
# ============================
#
# These statements execute immediately when the module is run.
# No explicit main() function is required.
#

tests = [
    ([1, 1, 1], 2, 2),                 # expected 2: [1,1] twice
    ([1, 2, 3], 3, 2),                 # expected 2: [1,2], [3]
    ([3, 4, 7, 2, -3, 1, 4, 2], 7, 4), # expected 4
    ([], 0, 0),                        # expected 0 (empty input)
]

print("=== Test: subarray_sum ===\n")

# Run each test case and compare the result with the expected value
for numbers, k_value, expected in tests:
    result = subarray_sum(numbers, k_value)
    print(f"Input: nums={numbers}, k={k_value}")
    print(f"Result: {result} (expected {expected})\n")

File: code_samples/section7/problems/problem7_4/problem7_4.cs
/**
 * SubarraySum
 *
 * Purpose:
 *   Counts the number of contiguous subarrays whose elements sum to k.
 *   This is the classic "Subarray Sum Equals K" problem.
 *
 * Parameters:
 *   nums - Input array of integers (may be empty; may include negative values).
 *   k    - Target sum.
 *
 * Returns:
 *   Number of contiguous subarrays whose sum is exactly k.
 *
 * Core idea (prefix sums + frequency dictionary):
 *   Let prefix be the running sum of nums[0..i].
 *   The sum of a subarray (j+1..i) is:
 *
 *       prefix[i] - prefix[j]
 *
 *   We want prefix[i] - prefix[j] == k
 *       => prefix[j] == prefix[i] - k
 *
 *   So for each position i:
 *     1) Compute current prefix sum P
 *     2) Add how many times (P - k) has appeared as a prefix sum before
 *     3) Record that we've seen prefix sum P one more time
 *
 * Why we seed freq[0] = 1:
 *   - Represents the "empty prefix" before reading any elements.
 *   - Allows subarrays starting at index 0 (where prefix == k) to be counted.
 *
 * Complexity (expected / average):
 *   - Time:  O(n)
 *       Each iteration performs O(1) average dictionary operations.
 *   - Space: O(n)
 *       In the worst case, every prefix sum is distinct.
 *
 * Notes:
 *   - Works with negative numbers (unlike many sliding-window approaches).
 *   - If sums might overflow int for large inputs, consider using long for prefix
 *     and Dictionary<long, int>.
 */
static int SubarraySum(int[] nums, int k) {
    // Frequency map: prefixSum -> number of times we've seen that prefix sum
    var freq = new Dictionary<int, int>
    {
        // Base case: prefix sum 0 occurs once before processing elements
        [0] = 1
    };

    int prefix = 0; // running prefix sum
    int count = 0;  // total number of subarrays with sum k

    // Process each element once
    foreach (var x in nums) {
        // Update running sum
        prefix += x;

        // We need prior prefix sums equal to (prefix - k)
        int need = prefix - k;

        // If such prefix sums exist, each occurrence forms a subarray ending here
        if (freq.TryGetValue(need, out int c)) {
            count += c;
        }

        // Record that we've seen this prefix sum.
        // If prefix already exists, increment its frequency; otherwise set to 1.
        freq[prefix] = freq.TryGetValue(prefix, out int cur) ? cur + 1 : 1;
    }

    return count;
}

// ============================
// Top-level tests
// ============================
//
// These are top-level statements (C# 9+). The array literals below use
// collection expressions (C# 12+).
//

int[][] tests =
[
    [1, 1, 1],                      // expected 2
    [1, 2, 3],                      // expected 2: [1,2], [3]
    [3, 4, 7, 2, -3, 1, 4, 2],      // expected 4
    []                              // expected 0
];

// Target k values aligned with tests[]
int[] kVals = [2, 3, 7, 0];

// Expected results aligned with tests[]
int[] expected = [2, 2, 4, 0];

Console.WriteLine("=== Test: SubarraySum ===\n");

// Run each test case and compare to expected output
for (int i = 0; i < tests.Length; i++)
{
    int[] nums = tests[i];
    int k = kVals[i];

    Console.WriteLine($"Input: [{string.Join(", ", nums)}], k = {k}");

    // Call the function under test
    int result = SubarraySum(nums, k);

    Console.WriteLine($"Result: {result} (expected {expected[i]})\n");
}
